ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/table/LoadPartitions.java

```diff
@@ -0,0 +1,259 @@
+      TableEvent event, String dbNameToLoadIn, String lastPartitionReplicated)
+      throws HiveException, IOException {
+    this.tracker = new TaskTracker(tableTracker);
+    this.event = event;
+    this.context = context;
+    this.lastPartitionReplicated = lastPartitionReplicated;
+    this.tableContext = tableContext;
+
+    this.tableDesc = tableContext.overrideProperties(event.tableDesc(dbNameToLoadIn));
+    this.table = ImportSemanticAnalyzer.tableIfExists(tableDesc, context.hiveDb);
+  }
+
+  private String location() throws MetaException, HiveException {
+    Database parentDb = context.hiveDb.getDatabase(tableDesc.getDatabaseName());
+    if (!tableContext.waitOnPrecursor()) {
+      return context.warehouse.getDefaultTablePath(parentDb, tableDesc.getTableName()).toString();
+    } else {
+      Path tablePath = new Path(
+          context.warehouse.getDefaultDatabasePath(tableDesc.getDatabaseName()),
+          MetaStoreUtils.encodeTableName(tableDesc.getTableName().toLowerCase())
+      );
+      return context.warehouse.getDnsPath(tablePath).toString();
+    }
+  }
+
+  public TaskTracker tasks() throws SemanticException {
+    try {
+      /*
+      We are doing this both in load table and load partitions
+       */
+      if (tableDesc.getLocation() == null) {
+        tableDesc.setLocation(location());
+      }
+
+      if (table == null) {
+        //new table
+
+        table = new Table(tableDesc.getDatabaseName(), tableDesc.getTableName());
+        if (isPartitioned(tableDesc)) {
+          updateReplicationState(initialReplicationState());
+          return forNewTable();
+        }
+      } else {
+        // existing
+
+        if (table.isPartitioned()) {
+          List<AddPartitionDesc> partitionDescs = event.partitionDescriptions(tableDesc);
+          if (!event.replicationSpec().isMetadataOnly() && !partitionDescs.isEmpty()) {
+            updateReplicationState(initialReplicationState());
+            return forExistingTable(lastPartitionReplicated);
+          }
+        }
+      }
+      return tracker;
+    } catch (Exception e) {
+      throw new SemanticException(e);
+    }
+  }
+
+  private void updateReplicationState(ReplicationState replicationState) throws SemanticException {
+    if (!tracker.canAddMoreTasks()) {
+      tracker.setReplicationState(replicationState);
+    }
+  }
+
+  private ReplicationState initialReplicationState() throws SemanticException {
+    return new ReplicationState(new PartitionState(tableDesc.getTableName(), lastPartitionReplicated));
+  }
+
+  private TaskTracker forNewTable() throws Exception {
+    Iterator<AddPartitionDesc> iterator = event.partitionDescriptions(tableDesc).iterator();
+    while (iterator.hasNext() && tracker.canAddMoreTasks()) {
+      AddPartitionDesc addPartitionDesc = iterator.next();
+      tracker.addTask(addSinglePartition(table, addPartitionDesc));
+      ReplicationState currentReplicationState =
```

Comment:
Instead of creating currentReplicationState object for each partition, it is enough to retain addPartitionDesc of last iteration and set it outside only once.